File: termin/visualization/core/entity_registry.py
"""
EntityRegistry - global registry for entity lookup by UUID and pick_id.

Singleton pattern similar to ResourceManager.
Entities register themselves on creation and unregister on removal.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Optional
from weakref import WeakValueDictionary

if TYPE_CHECKING:
    from termin.visualization.core.entity import Entity


class EntityRegistry:
    """
    Global registry for entity lookup by UUID and pick_id.

    Uses WeakValueDictionary for UUID lookup to avoid preventing garbage collection.
    Uses regular dict for pick_id lookup (pick_id is lazily assigned).
    """

    _instance: Optional["EntityRegistry"] = None

    def __init__(self):
        self._entities_by_uuid: WeakValueDictionary[str, "Entity"] = WeakValueDictionary()
        self._entities_by_pick_id: dict[int, "Entity"] = {}

    @classmethod
    def instance(cls) -> "EntityRegistry":
        """Get singleton instance."""
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    @classmethod
    def reset(cls) -> None:
        """Reset singleton (for testing)."""
        cls._instance = None

    def register(self, entity: "Entity") -> None:
        """Register entity by UUID."""
        if entity.uuid:
            self._entities_by_uuid[entity.uuid] = entity

    def register_pick_id(self, pick_id: int, entity: "Entity") -> None:
        """Register entity by pick_id."""
        self._entities_by_pick_id[pick_id] = entity

    def unregister(self, entity: "Entity") -> None:
        """Unregister entity from UUID registry."""
        if entity.uuid and entity.uuid in self._entities_by_uuid:
            del self._entities_by_uuid[entity.uuid]

    def unregister_pick_id(self, pick_id: int) -> None:
        """Unregister entity from pick_id registry."""
        self._entities_by_pick_id.pop(pick_id, None)

    def get(self, uuid: str) -> Optional["Entity"]:
        """Get entity by UUID."""
        return self._entities_by_uuid.get(uuid)

    def get_by_pick_id(self, pick_id: int) -> Optional["Entity"]:
        """Get entity by pick_id."""
        return self._entities_by_pick_id.get(pick_id)

    def clear(self) -> None:
        """Clear all registered entities."""
        self._entities_by_uuid.clear()
        self._entities_by_pick_id.clear()

    def swap_registries(
        self,
        new_by_uuid: WeakValueDictionary[str, "Entity"],
        new_by_pick_id: dict[int, "Entity"],
    ) -> tuple[WeakValueDictionary[str, "Entity"], dict[int, "Entity"]]:
        """
        Swap registries and return old ones.
        Used for game mode isolation.
        """
        old_by_uuid = self._entities_by_uuid
        old_by_pick_id = self._entities_by_pick_id
        self._entities_by_uuid = new_by_uuid
        self._entities_by_pick_id = new_by_pick_id
        return old_by_uuid, old_by_pick_id


__all__ = ["EntityRegistry"]

File: termin-app/cpp/termin/entity/entity_registry.cpp
#include "entity_registry.hpp"

namespace termin {

EntityRegistry& EntityRegistry::instance() {
    static EntityRegistry inst;
    return inst;
}

void EntityRegistry::register_entity(const Entity& entity) {
    if (!entity.valid()) return;

    const char* u = entity.uuid();
    if (u && u[0]) {
        by_uuid_[u] = entity;
    }

    // Also register by pick_id
    uint32_t pid = entity.pick_id();
    if (pid != 0) {
        by_pick_id_[pid] = entity;
    }
}

void EntityRegistry::unregister_entity(const Entity& entity) {
    if (!entity.valid()) return;

    const char* u = entity.uuid();
    if (u && u[0]) {
        by_uuid_.erase(u);
    }

    // Also remove from pick_id registry if registered
    for (auto it = by_pick_id_.begin(); it != by_pick_id_.end(); ) {
        if (it->second == entity) {
            it = by_pick_id_.erase(it);
        } else {
            ++it;
        }
    }
}

Entity EntityRegistry::get(const std::string& uuid) const {
    auto it = by_uuid_.find(uuid);
    return (it != by_uuid_.end()) ? it->second : Entity();
}

void EntityRegistry::register_pick_id(uint32_t pick_id, const Entity& entity) {
    if (entity.valid()) {
        by_pick_id_[pick_id] = entity;
    }
}

void EntityRegistry::unregister_pick_id(uint32_t pick_id) {
    by_pick_id_.erase(pick_id);
}

Entity EntityRegistry::get_by_pick_id(uint32_t pick_id) const {
    auto it = by_pick_id_.find(pick_id);
    return (it != by_pick_id_.end()) ? it->second : Entity();
}

void EntityRegistry::clear() {
    by_uuid_.clear();
    by_pick_id_.clear();
}

std::pair<std::unordered_map<std::string, Entity>,
          std::unordered_map<uint32_t, Entity>>
EntityRegistry::swap_registries(std::unordered_map<std::string, Entity> new_by_uuid,
                                std::unordered_map<uint32_t, Entity> new_by_pick_id) {
    auto old_by_uuid = std::move(by_uuid_);
    auto old_by_pick_id = std::move(by_pick_id_);

    by_uuid_ = std::move(new_by_uuid);
    by_pick_id_ = std::move(new_by_pick_id);

    return {std::move(old_by_uuid), std::move(old_by_pick_id)};
}

} // namespace termin
